### DISCUSSION TITLE
Lorentz autofit code - trouble getting accurate fit

### DISCUSSION THREAD

**Discussion by Nikolaos-Matthaiakakis** [NanoComp/meep#2561] [Q&A]
Dear all,

I am trying to use the script provided at [docs](https://meep.readthedocs.io/en/latest/Materials/#how-do-i-import-n-and-k-values-into-meep ), [script](https://github.com/NanoComp/meep/blob/master/python/examples/eps_fit_lorentzian.py ) to autofit the refractive index of [PDMS](https://refractiveindex.info/?shelf=organic&book=polydimethylsiloxane&page=Zhang-5-1) in the infrared spectral range.

I have used many different combinations of Lorentzian numbers (num_lorentzians) and iterations (num_repeat) but the script seems to only fit one of the Lorentzian peaks that appear in the experimental results.

Is there any advice for improving the fitting process?

![eps_fit_sample](https://github.com/NanoComp/meep/assets/58590846/00499d0a-2324-4e30-8d38-205d1baf766a)

Thank you for your time

**Comment by oskooi** [comment]
In my experience, a key parameter for obtaining decent fits is `eps_inf`:

https://github.com/NanoComp/meep/blob/d23e3427b19cfdb32b5cad3024f42ba8ec132aae/python/examples/eps_fit_lorentzian.py#L75-L80

Small changes in `eps_inf` tend to be produce large changes in the final results. You can try manually tuning this parameter to see this effect for yourself.

**Reply by Nikolaos-Matthaiakakis** [reply]
Thank you

I tried the following and many other options but the code seems to simply focus on plotting this one peak. I have not made any changes to the code other than for the changes in these values so I am not sure why this happens.
```
eps_inf = 1  # should be >= 1.0 for stability and chosen such that np.amin(np.real(eps)) is ~1.0
num_lorentzians = 12  
num_repeat = (
    30  # number of times to repeat local optimization with random initial values
)
```
Result:
![eps_fit_sample](https://github.com/NanoComp/meep/assets/58590846/9835db1b-26c9-46a9-9d21-ca7834cd392d)


```
eps_inf = 1.5  # should be >= 1.0 for stability and chosen such that np.amin(np.real(eps)) is ~1.0
num_lorentzians = 12  
num_repeat = (
    30  # number of times to repeat local optimization with random initial values
)
```
Result:
![eps_fit_sample](https://github.com/NanoComp/meep/assets/58590846/2d0abac1-f269-45ca-a8ce-a76a5eaf58b4)

```
eps_inf = 2 # should be >= 1.0 for stability and chosen such that np.amin(np.real(eps)) is ~1.0
num_lorentzians = 12  
num_repeat = (
    30  # number of times to repeat local optimization with random initial values
)
```
Result:
![eps_fit_sample](https://github.com/NanoComp/meep/assets/58590846/23ff0aac-d691-4ec7-a8c4-ba66ae754551)

**Reply by smartalecH** [reply]
How are you plotting things?

Also, how well does the optimization converge?

**Reply by Nikolaos-Matthaiakakis** [reply]
I simply ran the code that can be found in the links of my original post (from the meep documentation). This code includes the plots so the only thing I changed was the metarial csv file.

**Comment by oskooi** [comment]
As an additional check, note that the script is fitting $\varepsilon(f)$ where $f$ is the frequency. You therefore need to be consistent with converting your input data from wavelength to frequency and then converting the fit $\overline{\varepsilon}$ from frequency to wavelength when comparing the input data to the fit using a plot: $\varepsilon(\lambda) \rightarrow \varepsilon(f) \rightarrow \overline{\varepsilon}(f) \rightarrow \overline{\varepsilon}(\lambda)$.

https://github.com/NanoComp/meep/blob/95ddacdda9754b0c6613b2a25a1154e87ab427e9/python/examples/eps_fit_lorentzian.py#L88

**Reply by Nikolaos-Matthaiakakis** [reply]
Thank you for the reply. I simply use the code found in the meep documentation as is and load my material data as specified in the comments  and only change the eps_inf, num_lorentzians, and num_repeat values.

I see that the code already handles the conversions after loading the wavelengths in nm and the final plots also show that the correct values have been used.

```
# format of input data file is three comma-separated columns:
# wavelength (nm), real(n), complex(n)
mydata = np.genfromtxt("mymaterial.csv", delimiter=",") # https://refractiveindex.info/?shelf=organic&book=polydimethylsiloxane&page=Zhang-5-1
n = mydata[:, 1] + 1j * mydata[:, 2]
eps_inf = 1.5  # should be >= 1.0 for stability and chosen such that np.amin(np.real(eps)) is ~1.0
eps = np.square(n) - eps_inf
wl = mydata[:, 0]                *10**3 # my wavelength data was in um so I convert it to nm here
wl_min = 400  # minimum wavelength (units of nm)
wl_max = 19442  # maximum wavelength (units of nm)
start_idx = np.where(wl > wl_min)
idx_start = start_idx[0][0]
end_idx = np.where(wl < wl_max)
idx_end = end_idx[0][-1] + 1
freqs = 1000 / wl  # units of 1/μm
freqs_reduced = freqs[idx_start:idx_end]
wl_reduced = wl[idx_start:idx_end]
eps_reduced = eps[idx_start:idx_end]

num_lorentzians = 12
num_repeat = (
    30  # number of times to repeat local optimization with random initial values
```

I got the fitting for PDMS manually but I do want to have this working for the future so I will investigate again.

**Comment by ramji28** [comment]
@Nikolaos-Matthaiakakis Have you found a solution to this problem?

**Reply by Nikolaos-Matthaiakakis** [reply]
Sorry I never investigated the issue again since I posted this.
